### PR TITLE
Acme attestation format

### PR DESCRIPTION
**Pull request by maraino**
### Description

This PR adds the attestation format flags to configure the ACME provisioners.

Notice that there's still an open discussion about the name of this feature. Should it be `--platform` or the more technical name `--attestation-format`. See https://github.com/smallstep/linkedca/pull/27.

Depends on:
- [x] https://github.com/smallstep/certificates/pull/1048
- [x] https://github.com/smallstep/linkedca/pull/27

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -19,7 +19,11 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [Unreleased]
 ### Added
+- Added support for configuring ACME devices-attest-01 challenges.
+- Added support for ACME device-attest-01 challenges with YubiKeys.
 - Added support for SUSE13 and upwards for `step certificate install`
```
Comment (dopey) [inline CHANGELOG.md:24]:
```suggestion
- Added support for SUSE13 and upwards for `step certificate install`.
```

File: command/ca/provisioner/add.go [hunk 2]
```diff
@@ -803,6 +811,26 @@ func validateChallengeFlag(ctx *cli.Context, typ linkedca.Provisioner_Type) erro
...
+			case "apple", "step", "tpm":
+			default:
+				return errs.InvalidFlagValue(ctx, "attestation-format", v, "apple, step and tpm")
+			}
+		}
+		for _, v := range ctx.StringSlice("remove-attestation-format") {
+			switch strings.ToLower(v) {
+			case "apple", "step", "tpm":
+			default:
+				return errs.InvalidFlagValue(ctx, "remove-attestation-format", v, "apple, step and tpm")
```
Comment (dopey) [inline command/ca/provisioner/add.go:827]:
```suggestion
				return errs.InvalidFlagValue(ctx, "remove-attestation-format", v, "apple, step, and tpm")
```

File: command/ca/provisioner/add.go [hunk 3]
```diff
@@ -803,6 +811,26 @@ func validateChallengeFlag(ctx *cli.Context, typ linkedca.Provisioner_Type) erro
 	return nil
 }
 
+func validateAttestationFormatFlag(ctx *cli.Context, typ linkedca.Provisioner_Type) error {
+	if typ == linkedca.Provisioner_ACME {
+		for _, v := range ctx.StringSlice("attestation-format") {
+			switch strings.ToLower(v) {
+			case "apple", "step", "tpm":
+			default:
+				return errs.InvalidFlagValue(ctx, "attestation-format", v, "apple, step and tpm")
```
[the remark this answers is not in the document]
Comment (dopey) [reply command/ca/provisioner/add.go:820]: oxford comma

File: command/ca/provisioner/add.go [hunk 4]
```diff
@@ -815,10 +843,28 @@ func acmeChallengeToLinkedca(challenges []string) []linkedca.ACMEProvisioner_Cha
...
-			ret = append(ret, linkedca.ACMEProvisioner_TLS_ALPN_O1)
+			ret = append(ret, linkedca.ACMEProvisioner_TLS_ALPN_01)
 		case "device-attest-01":
 			ret = append(ret, linkedca.ACMEProvisioner_DEVICE_ATTEST_01)
 		}
 	}
 	return ret
 }
+
+// acmeAttestationFormatToLinkedca returns the linkedca attestation format types
```
Comment (dopey) [inline command/ca/provisioner/add.go:854]:
```
// acmeAttestationFormatToLinkedca returns the linkedca attestation format types
// for the attestation-format flag. This function assumes the options have been 
// previously validated.
```

File: command/ca/provisioner/provisioner.go [hunk 5]
```diff
@@ -368,6 +368,32 @@ Use the flag multiple times to remove multiple challenges.`,
...
+		Usage: `Enable an ACME attestation statement <format> in the provisioner. Use the flag
+multiple times to configure multiple challenges.
+
+The supported ACME attestation formats are:
+
+**apple**
+: With the apple format, Apple devices can use the device-attest-01 challenge to
+get a new certificate.
+
+**step**
```
Comment (dopey) [inline command/ca/provisioner/provisioner.go:382]: Why do we call this `step`?

File: command/ca/provisioner/update.go [hunk 6]
```diff
@@ -609,6 +612,17 @@ func updateACMEDetails(ctx *cli.Context, p *linkedca.Provisioner) error {
 	if ctx.IsSet("challenge") || ctx.IsSet("remove-challenge") {
 		details.Challenges = sliceutil.RemoveDuplicates(details.Challenges)
 	}
+	if ctx.IsSet("remove-attestation-format") {
```
Comment (dopey) [inline command/ca/provisioner/update.go:615]: Shouldn't we be validating these? like we do for provisioner add? Same question about the `challenge` flag.

**Review by dopey — CHANGES_REQUESTED** [review]
A couple questions and a few recommended changes in documentation.

File: command/ca/provisioner/add.go (hunk 4 above)
Comment (maraino) [reply command/ca/provisioner/add.go:854]: Fixed with e4a62de

File: command/ca/provisioner/add.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (maraino) [reply command/ca/provisioner/add.go]: Fixed with e4a62de

File: command/ca/provisioner/add.go (hunk 2 above)
Comment (maraino) [reply command/ca/provisioner/add.go]: Fixed with e4a62de

File: command/ca/provisioner/provisioner.go (hunk 5 above)
Comment (maraino) [reply command/ca/provisioner/provisioner.go:382]: We need a custom format to differentiate the different requests, apple has its own. 

The format `step` is based on the attestation format `packed` defined in https://www.w3.org/TR/2019/REC-webauthn-1-20190304/#packed-attestation, but the verification of the attestation certificates is slightly different.

Comment (maraino) [reply command/ca/provisioner/provisioner.go:382]: We can switch to `packed` the format is exactly the same. Just the verification is different.

File: command/ca/provisioner/update.go (hunk 6 above)
Comment (maraino) [reply command/ca/provisioner/update.go:620]: Fixed with 021d372

**Review by dopey — APPROVED** [review]
lgtm

**Pull request merged**